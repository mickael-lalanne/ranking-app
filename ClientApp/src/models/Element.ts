export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
