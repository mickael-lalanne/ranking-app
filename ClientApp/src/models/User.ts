export type UserId = string | null | undefined;

export interface User {
    id: UserId;
}
