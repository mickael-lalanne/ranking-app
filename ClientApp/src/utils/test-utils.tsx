import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
