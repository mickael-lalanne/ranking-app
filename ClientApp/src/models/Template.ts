import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
