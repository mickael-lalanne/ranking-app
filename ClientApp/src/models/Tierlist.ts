import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
