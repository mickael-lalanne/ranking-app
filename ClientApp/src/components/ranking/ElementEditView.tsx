import React, { useState } from 'react';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import { css } from '@emotion/css';
import { Element } from '../../models/Element';
import { generateRandomId } from '../../services/Util';
import { EEditViewMode } from '../../models/Template';

/**
 * View displayed when the user wants to create a new element or edit an existing one
 */
const ElementEditView = ({createCallback, cancelCallback, editViewMode}: {
    createCallback: (element: Element) => void,
    cancelCallback: () => void,
    editViewMode: EEditViewMode
}) => {
    const [elementName, setElementName] = useState<string>();

    // Called when the "Create" button of the edit view has been clicked
    const createElement = () => {
        createCallback({
            id: generateRandomId(),
            name: elementName as string,
            image: 'tmp_img'
        });
    };

    // Called when the element name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setElementName(e.target.value);
    };

    if (editViewMode === EEditViewMode.EditElement) {
        return (<div className={element_edit_view_container_style}>
            <TextField
                label="Element name"
                variant="outlined"
                onChange={onNameFieldChange}
            />
            <div className={element_image_style}>
                <div>Image :</div>
                <div>Todo...</div>
            </div>
            <div className={footer_buttons_style}>
                <div className="app_spacer"></div>
                <Button variant="outlined" onClick={cancelCallback}>Cancel</Button>
                <Button variant="contained" onClick={createElement} style={{ marginLeft: '10px' }}>Create</Button>
            </div>
        </div>);
    }
    return <></>;
};

export default ElementEditView;

/**
 * CSS STYLES
 */
const element_edit_view_container_style = css({
    display: 'flex',
    flexDirection: 'column',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px;',
    marginTop: '15px',
    padding: '30px'
});

const footer_buttons_style = css({
    display: 'flex'
});

const element_image_style = css({
    display: 'flex',
    alignItems: 'center',
    marginTop: '10px'
})
