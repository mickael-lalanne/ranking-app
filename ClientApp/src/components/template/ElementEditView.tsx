import React, { useState, useEffect } from 'react';
import TextField from '@mui/material/TextField';
import Button from '@mui/material/Button';
import Alert from '@mui/material/Alert';
import { css } from '@emotion/css';
import { Element } from '../../models/Element';
import { ResizedImage, generateRandomId, resizeImage } from '../../services/Util';
import { EEditViewMode } from '../../models/Template';
import UploadFileIcon from '@mui/icons-material/UploadFile';
import ElementPreview from '../shared/ElementPreview';
import AddElementButton from '../shared/AddElementButton';
import { ELEMENT_SIZE } from '../../utils/css-utils';
import { useAppSelector } from '../../app/hooks';

/**
 * View displayed when the user wants to create a new element or edit an existing one
 */
const ElementEditView = ({createCallback, cancelCallback, editViewMode, defaultImages}: {
    createCallback: (elements: Element[]) => void,
    cancelCallback: () => void,
    editViewMode: EEditViewMode,
    defaultImages: ResizedImage[]
}) => {
    const [currentElements, setCurrentElements] = useState<Element[]>([]);
    const [uploadBtnHover, setUploadBtnHover] = useState<string>();

    /**
     * Called when the defaultImages props has changed
     * Set the current elements with thanks to default images data
     */
    useEffect(() => {
        const elements: Element[] = _convertImagesToElements(defaultImages);
        setCurrentElements(elements);
    }, [defaultImages]);

    const loading: boolean = useAppSelector(state => state.application.loading);

    // Called when the "Create" button of the edit view has been clicked
    const createElements = () => {
        createCallback(currentElements);
    };

    // Called when the element name has changed
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>, elementId: string) => {
        const eltToUpdateIndex: number = currentElements.findIndex(e => e.id === elementId);
        if (eltToUpdateIndex > -1) {
            currentElements[eltToUpdateIndex].name = e.target.value;
            setCurrentElements([...currentElements]);
        }
    };

    // Called when the user has chosen an image with the file input
    const onImageFieldChange = async (
        e: React.ChangeEvent<HTMLInputElement>,
        elementId: string
    ): Promise<void> => {
        const eltToUpdateIndex: number = currentElements.findIndex(e => e.id === elementId);
        if (eltToUpdateIndex > -1) {
            if (e.target.files) {
                const resizedImage = await resizeImage(e.target.files[0]);
                // Be sure to create a new object so ElementPreview can detect the change
                // Otherwise the useEffect for the element props is not called
                // So we need to create a new reference (for objects AND arrays)
                currentElements[eltToUpdateIndex] = {
                    ...currentElements[eltToUpdateIndex],
                    image: resizedImage.source
                };
                setCurrentElements([...currentElements]);
            }
        }
        setUploadBtnHover(undefined);
    };

    // Called when the cancel button has been clicked
    const onCancelButtonClick = (): void => {
        // Reset data
        setCurrentElements([]);

        cancelCallback();
    };

    // Called when the user has selected one or many images for its elements
    const onElementImageInputChange = async (
        e: React.ChangeEvent<HTMLInputElement>
    ): Promise<void> => {
        if (e.target.files) {
            const resizeFilePromises: Promise<ResizedImage>[] = [];

            // Resize all images selectionned by the user
            Array.from(e.target.files).forEach(file => {
                resizeFilePromises.push(resizeImage(file));
            });

            // Once all images have been resized
            const resizedImages: ResizedImage[] = await Promise.all(resizeFilePromises);

            const newElements = _convertImagesToElements(resizedImages);
            setCurrentElements(currentElements.concat(newElements));
        }
    };

    // The content of the upload button
    // It can be the element's image if it exists, and the upload icon if hovered
    const UploadButtonContent = (elt: Element): React.JSX.Element => {
        const uploadIcon: React.JSX.Element = <UploadFileIcon className={upload_icon_style} />;

        return (<>
            <ElementPreview element={elt} readonly />
            {/* Show the upload icon only on hover */}
            { uploadBtnHover === elt.id ? uploadIcon : null }
        </>);
    };

    const ElementsEdition = (): React.JSX.Element[] => {
        const editionsView: React.JSX.Element[] = [];
        currentElements.forEach(elt => {
            const eltEditView = <div className={element_edit_view_content_style} key={elt.id}>
                <div className={element_image_style}>
                    <Button
                        variant="contained"
                        component="label"
                        className={upload_btn_style}
                        onMouseEnter={() => setUploadBtnHover(elt.id)}
                        onMouseLeave={() => setUploadBtnHover('')}
                    >
                        {UploadButtonContent(elt)}
                        <input
                            type="file"
                            className={image_input_style}
                            onChange={e => onImageFieldChange(e, elt.id!)}
                        ></input>
                    </Button>
                </div>
                <TextField
                    label="Element name"
                    variant="outlined"
                    style={{ flex: 1 }}
                    value={elt.name}
                    onChange={e => onNameFieldChange(e, elt.id!)}
                />
            </div>;

            editionsView.push(eltEditView);
        });

        return editionsView;
    };

    if (editViewMode === EEditViewMode.EditElement && !loading) {
        return (<div className={element_edit_view_container_style}>
            <Alert severity="info" className={warning_message_style}>
                FYI : all your images are automatically resized to have a maximum size of 150x150.
            </Alert>

            {ElementsEdition()}

            <div className={footer_style}>
                <AddElementButton changeCallback={onElementImageInputChange} />
                <div className="app_spacer"></div>
                <Button
                    variant="outlined"
                    onClick={onCancelButtonClick}
                    className={footer_btn_style}
                >
                        Cancel
                </Button>
                <Button
                    variant="contained"
                    onClick={createElements}
                    className={footer_btn_style}
                    style={{ marginLeft: '10px' }}
                    data-cy="create-element-button"
                >
                    Create
                </Button>
            </div>
        </div>);
    }
    return <></>;
};


/**
 * Convert images files into elements
 * @param {string[]} imagesFile the images (as DataURI) to convert
 * @returns {Element[]} element array with a random id and the image's name as name
 */
const _convertImagesToElements = (images: ResizedImage[]): Element[] => {
    const elements: Element[] = images.map(img => {
        return {
            id: generateRandomId(),
            image: img.source,
            name: img.name
        };
    });

    return elements;
}


export default ElementEditView;

/**
 * CSS STYLES
 */
const element_edit_view_container_style = css({
    display: 'flex',
    flexDirection: 'column',
    boxShadow: 'rgba(149, 157, 165, 0.2) 0px 8px 24px;',
    marginTop: '15px',
    padding: '30px'
});

const element_edit_view_content_style = css({
    display: 'flex',
    alignItems: 'center',
    marginBottom: '15px'
});

const footer_style = css({
    display: 'flex'
});

const footer_btn_style = css({
    marginTop: '10px !important',
    marginBottom: '10px !important'
});

const element_image_style = css({
    position: 'relative',
    display: 'flex',
    alignItems: 'center',
    marginRight: '10px',
    width: ELEMENT_SIZE,
    height: ELEMENT_SIZE
});

const upload_btn_style = css({
    padding: '0px !important',
    height: '100%',
    minWidth: 'unset !important',
    width: '100%'
});

const upload_icon_style = css({
    position: 'absolute',
    width: '100% !important',
    height: '100% !important',
    top: 0,
    left: 0,
    padding: '7.5px',
    color: 'white'
});

const image_input_style = css({
    opacity: 0,
    zIndex: -9999,
});

const warning_message_style = css({
    marginBottom: '25px'    
});
