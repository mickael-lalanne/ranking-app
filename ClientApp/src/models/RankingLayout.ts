import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
