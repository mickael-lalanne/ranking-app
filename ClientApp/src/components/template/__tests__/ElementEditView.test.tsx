import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import { screen } from '@testing-library/react';
import ElementEditView, { ElementEditViewProps } from '../ElementEditView';
import { EEditViewMode } from '../../../models/Template';
import userEvent from '@testing-library/user-event';
import { Element } from '../../../models/Element';

const DEFAULT_PROPS: ElementEditViewProps = {
    createCallback: jest.fn(),
    cancelCallback: jest.fn(),
    editViewMode: EEditViewMode.EditElement,
    defaultImages: [{ source: 'img-source', name: 'img-name' }],
};

const ElementEditViewTest = (props: ElementEditViewProps) => {
    return (
        <ElementEditView
            createCallback={props.createCallback}
            cancelCallback={props.cancelCallback}
            editViewMode={props.editViewMode}
            defaultImages={props.defaultImages}
        />
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(
            ElementEditViewTest(DEFAULT_PROPS)
        );

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    beforeEach(() => {
        (DEFAULT_PROPS.createCallback as jest.Mock).mockClear();
        (DEFAULT_PROPS.cancelCallback as jest.Mock).mockClear();
    });

    it('creates element', async () => {
        renderWithProviders(ElementEditViewTest(DEFAULT_PROPS));

        // Change the element name
        const elementName = 'My element name';
        await userEvent.clear(
            screen.getByDisplayValue(DEFAULT_PROPS.defaultImages[0].name)
        );
        await userEvent.type(
            screen.getByLabelText('Element name'),
            elementName
        );

        // Click on the create button
        await userEvent.click(screen.getByTestId('create-element-button'));
        const createCall: Element[] = (
            DEFAULT_PROPS.createCallback as jest.Mock
        ).mock.calls[0][0];
        expect(createCall[0].image).toBe(DEFAULT_PROPS.defaultImages[0].source);
        expect(createCall[0].name).toBe(elementName);
    });

    it('cancels element', async () => {
        renderWithProviders(ElementEditViewTest(DEFAULT_PROPS));

        await userEvent.click(screen.getByTestId('cancel-element-button'));
        expect(DEFAULT_PROPS.cancelCallback).toHaveBeenCalledTimes(1);
    });
});
