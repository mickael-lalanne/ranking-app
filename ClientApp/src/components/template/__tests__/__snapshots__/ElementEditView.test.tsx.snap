// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <div
      class="css-8jos58"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardInfo MuiAlert-standard css-d2znx6 css-1fuz9qr-MuiPaper-root-MuiAlert-root"
        role="alert"
      >
        <div
          class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
            data-testid="InfoOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
        >
          FYI : all your images are automatically resized to have a maximum size of 150x150.
        </div>
      </div>
      <div
        class="css-14p7st7"
      >
        <div
          class="css-on4bb6"
        >
          <label
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1bu6aj css-sghohy-MuiButtonBase-root-MuiButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="css-13ambje"
              data-cy="element-preview"
              data-testid="element-preview"
              draggable="false"
              style="height: 75px; width: 75px; opacity: 1;"
            >
              <img
                class="css-40zi5q"
                data-cy="element-preview-img"
                src="https://res.cloudinary.com/ranking-app/image/upload/img-source"
              />
            </div>
            <input
              class="css-1y5dr88"
              type="file"
            />
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </label>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          style="flex: 1;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for=":r0:"
            id=":r0:-label"
          >
            Element name
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value="img-name"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Element name
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="css-k008qs"
      >
        <label
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-13ambje css-1t5b4kh css-sghohy-MuiButtonBase-root-MuiButton-root"
          role="button"
          tabindex="0"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddCircleOutlineIcon"
            focusable="false"
            style="width: 40px; height: 40px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
          <input
            accept="image/*"
            class="css-195j79u"
            data-cy="add-element-input"
            data-testid="add-element-input"
            multiple=""
            type="file"
          />
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </label>
        <div
          class="app_spacer"
        />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-2pqoxq css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="cancel-element-button"
          tabindex="0"
          type="button"
        >
          Cancel
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-2pqoxq css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-cy="create-element-button"
          data-testid="create-element-button"
          style="margin-left: 10px;"
          tabindex="0"
          type="button"
        >
          Create
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
